src/addrman.h

```diff
@@ -534,137 +553,133 @@ friend class CAddrManTest;
     }
 
     //! Return the number of (unique) addresses in all tables.
-    size_t size() const
+    size_t size() const EXCLUSIVE_LOCKS_REQUIRED(!m_addrman_mutex)
     {
-        LOCK(cs); // TODO: Cache this in an atomic to avoid this overhead
-        return vRandom.size();
+        LOCK(m_addrman_mutex); // TODO: Cache this in an atomic to avoid this overhead
+        return sizeNonLockerHelper();
```

Comment:
It would be nice to implement that TODO comment because it would make it possible to avoid any mutex requirements from `size()`.

I did so in https://github.com/vasild/bitcoin/commit/f25d8a7c6de0e45b037f91138e1b4a72e7a1653f. If you like it you can replace the commit `refactor: Prevent double lock in CAddrMan::size()` with that one. In this branch https://github.com/vasild/bitcoin/tree/200610-addrman-mx that is done and also the subsequent commits are adjusted.

Or, if you don't like the caching, then maybe just delete the TODO comment.

Reply:
The commit you [suggested](https://github.com/vasild/bitcoin/commit/f25d8a7c6de0e45b037f91138e1b4a72e7a1653f) does change `CAddrMan::size()` behavior. So I'd prefer to keep this PR as clean refactoring with no behavior changes.

Reply:
ok